spec/client/ics-028-wasm-client/README.md

```diff
@@ -0,0 +1,451 @@
+
+### Definitions
+
+Functions & terms are as defined in [ICS 2](../ics-002-client-semantics).
+
+`currentTimestamp` is as defined in [ICS 24](../ics-024-host-requirements).
+
+`WASM Client Code` refers to wasm bytecode stored in the client store, which provides target blockchain specific implementation of [ICS 2](../ics-002-client-semantics).
+
+`WASM Client` refers to a particular instance of `WASM Client Code` defined as a tuple `(WASM Client Code, ClientID)`.
+
+`WASM VM` refers to a virtual machine capable of executing valid wasm bytecode.
+
+
+### Desired Properties
+
+This specification must satisfy the client interface defined in ICS 2.
+
+## Technical Specification
+
+This specification depends on correctness of the `WASM Client code` in context of consensus algorithm of its target `blockchain`, as well as correct instantiation of `WASM Client`.
+
+### Client state
+
+The wasm client state tracks location of the wasm bytecode via `codeId`. Binary data represented by `data` field is opaque and only interpreted by the WASM Client Code. `type` represents client type.
+`type` and `codeId` both are immutable.
+
+```typescript
+interface ClientState {
+  codeId: []byte
+  data: []byte
+  frozen: boolean
+  frozen_height: Height
+  latest_height: Height
+  type: String
+}
+```
+
+### Consensus state
+
+The WASM consensus state tracks the timestamp (block time), `WASM Client code` specific fields and commitment root for all previously verified consensus states.
+`type` and `codeId` both are immutable.
+
+```typescript
+interface ConsensusState {
+  codeId: []byte
+  data: []byte
+  timestamp: uint64
+  root: MerkleRoot
+  type: String
+}
+```
+
+### Height
+
+The height of a WASM light client instance consists of two `uint64`s: the epoch number, and the height in the epoch.
+
+```typescript
+interface Height {
+  epochNumber: uint64
+  epochHeight: uint64
+}
+```
+
+Comparison between heights is implemented as follows:
+
+```typescript
+function compare(a: Height, b: Height): Ord {
+  if (a.epochNumber < b.epochNumber)
+    return LT
+  else if (a.epochNumber === b.epochNumber)
+    if (a.epochHeight < b.epochHeight)
+      return LT
+    else if (a.epochHeight === b.epochHeight)
+      return EQ
+  return GT
+}
+```
+
+This is designed to allow the height to reset to `0` while the epoch number increases by one in order to preserve timeouts through zero-height upgrades.
+
+### Headers
+
+Contents of wasm client headers depend upon `WASM Client Code`.
+
+```typescript
+interface Header {
+  data: []byte
+}
+```
+
+### Misbehaviour
+
+The `Misbehaviour` type is used for detecting misbehaviour and freezing the client - to prevent further packet flow - if applicable.
+WASM client `Misbehaviour` consists of two headers at the same height both of which the light client would have considered valid.
+
+```typescript
+interface Misbehaviour {
```

Comment:
why can't this be up to the WASM code

Reply:
@cwgoes See my previous comment.

Reply:
See my previous response

Reply:
These are the methods without access to storage
## Misbehaviour ([link](https://github.com/cosmos/ibc-go/blob/main/modules/core/exported/client.go#L201)):

```
ClientType() string
GetClientID() string
ValidateBasic() error
```

Reply:
These are methods without access to the storage:

## Misbehaviour ([link](https://github.com/cosmos/ibc-go/blob/main/modules/core/exported/client.go#L201)):**
```
ClientType() string
GetClientID() string
ValidateBasic() error
```

This seems to be a stale definition:
```
interface Misbehaviour {
  fromHeight: Height  // doesn't seem to be used
  client_id: string, // missing, used by GetClientID()
  h1: Header 
  h2: Header
}
```

More correct version:
```
interface Misbehaviour {
  client_id: string,
  h1: Header 
  h2: Header
}
```

Reply:
With the two headers, we should be able to compare the headers for misbehavior without storage access, just code.

Reply:
yes, that sounds right

Reply:
> With the two headers, we should be able to compare the headers for misbehavior without storage access, just code.

Agreed.

Reply:
updated `Misbehaviour` [here](https://github.com/ParthDesai/ibc/commit/d7c33fc5c01ff4c07000f89ff9da04c31c27ed23#diff-081ac6bcbed0e39a406e9c83aa79a058605a646a4047d5be67cb84dd9b83877bL117)